fast_neural_style/neural_style/neural_style.py

```diff
@@ -0,0 +1,232 @@
+import argparse
+import os
+import sys
+import time
+
+import numpy as np
+import torch
+from torch.autograd import Variable
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from torchvision import datasets
+from torchvision import transforms
+
+import utils
+from transformer_net import TransformerNet
+from vgg import Vgg16
+
+
+def train(args):
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+
+    if args.cuda:
+        torch.cuda.manual_seed(args.seed)
+
+    transform = transforms.Compose([
+        transforms.Scale(args.image_size),
+        transforms.CenterCrop(args.image_size),
+        transforms.ToTensor(),
+        transforms.Lambda(lambda x: x.mul(255))
+    ])
+    train_dataset = datasets.ImageFolder(args.dataset, transform)
+    train_loader = DataLoader(train_dataset, batch_size=args.batch_size)
+
+    transformer = TransformerNet()
+    optimizer = Adam(transformer.parameters(), args.lr)
+    mse_loss = torch.nn.MSELoss()
+
+    vgg = Vgg16()
+    style_transform = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Lambda(lambda x: x.mul(255))
+    ])
+    style = utils.load_image(args.style_image, size=args.style_size)
+    style = style_transform(style)
+    style = style.repeat(args.batch_size, 1, 1, 1)
+
+    if args.cuda:
+        transformer.cuda()
+        vgg.cuda()
+        style = style.cuda()
+
+    style_v = Variable(style, volatile=True)
+    style_v = utils.normalize_batch(style_v)
+    features_style = vgg(style_v)
+    gram_style = [utils.gram_matrix(y).data for y in features_style]
+
+    for e in range(args.epochs):
+        transformer.train()
+        agg_content_loss = 0.
+        agg_style_loss = 0.
+        count = 0
+        for batch_id, (x, _) in enumerate(train_loader):
+            n_batch = len(x)
+            count += n_batch
+            optimizer.zero_grad()
+            x = Variable(x)
+            if args.cuda:
+                x = x.cuda()
+
+            y = transformer(x)
+
+            xc = Variable(x.data.clone(), volatile=True)
```

Comment:
I think that if you make all weights of the `vgg` network `requires_grad=False`, then it's not necessary to have inputs being `volatile` to have low memory usage. See [discussion here](https://pytorch.slackarchive.io/random/page-9/ts-1490356747249255). It would remove the need of always unwrapping / wrapping the variables for the losses, and would make the code much cleaner I think.